include/engine/api/match_api.hpp

```diff
@@ -115,17 +115,23 @@ class MatchAPI final : public RouteAPI
                 sub_matchings[matching_index.sub_matching_index].nodes[matching_index.point_index];
             auto waypoint = BaseAPI::MakeWaypoint(phantom);
             waypoint.values["matchings_index"] = matching_index.sub_matching_index;
+            waypoint.values["waypoint_index"] = matching_index.point_index;
             waypoint.values["alternatives_count"] =
                 sub_matchings[matching_index.sub_matching_index]
                     .alternatives_count[matching_index.point_index];
-            if (tidy_result.was_waypoint[trace_index])
+            // waypoint indices need to be adjusted if route legs were collapsed
+            // waypoint parameter assumes there is only one match object
+            if (!parameters.waypoints.empty())
             {
-                waypoint.values["waypoint_index"] = was_waypoint_idx;
-                was_waypoint_idx++;
-            }
-            else
-            {
-                waypoint.values["waypoint_index"] = util::json::Null();
+                if (tidy_result.was_waypoint[trace_index])
```

Comment:
In this case we should have an assertion here that there is only one sub-matching, since that is the assumption that is used here.